### PR TITLE
Example/feature engineering lambda

### PR DESCRIPTION
**Pull request by fernandonieuwveldt**
In this example we look at how we can create a full training and inference pipeline implemented only using the Keras library. As we build up our graph we also visualize our network. 

We will end up with only one artifact containing the full pipeline. This can easily be deployed and you do not need to create features with other libraries before feeding data to your model.

### CODE REVIEW

File: examples/structured_data/feature_engineering_with_lambda.py [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+    """check if age gt 50 and if gender is male"""
+    return tf.cast(tf.math.logical_and(x[0] > 50, x[1] == 1), dtype=tf.float32)
+
+
+def is_fixed(x):
+    """encode categoric feature if value is equal to fixed"""
+    return tf.cast(x == "fixed", dtype=tf.float32)
+
+
+def is_reversible(x):
```
Comment (fchollet) [inline examples/structured_data/feature_engineering_with_lambda.py:155]: Pretty mysterious

**Review by fchollet — COMMENTED** [review]
Thanks for the PR! Feature engineering for categorical data is a great topic. However, using `Lambda` layers is not recommended. They're not safely serializable and I wouldn't recommend them in production for this reason.

We also already have an example on structured data feature engineering here: https://keras.io/examples/structured_data/structured_data_classification_from_scratch/

I would recommend turning your example into a tutorial that focuses on something that's absent from the example above. Perhaps we could take the approach of doing the feature engineering in a single Layer subclass that takes in a dict of data. What do you think?

**Comment by fernandonieuwveldt** [comment]
> Thanks for the PR! Feature engineering for categorical data is a great topic. However, using `Lambda` layers is not recommended. They're not safely serializable and I wouldn't recommend them in production for this reason.
> 
> We also already have an example on structured data feature engineering here: https://keras.io/examples/structured_data/structured_data_classification_from_scratch/
> 
> I would recommend turning your example into a tutorial that focuses on something that's absent from the example above. Perhaps we could take the approach of doing the feature engineering in a single Layer subclass that takes in a dict of data. What do you think?

@fchollet That sounds great yes. I will have a look at changing it to subclass Layer class.

**Comment by fernandonieuwveldt** [comment]
> Thanks for the PR! Feature engineering for categorical data is a great topic. However, using `Lambda` layers is not recommended. They're not safely serializable and I wouldn't recommend them in production for this reason.
> 
> We also already have an example on structured data feature engineering here: https://keras.io/examples/structured_data/structured_data_classification_from_scratch/
> 
> I would recommend turning your example into a tutorial that focuses on something that's absent from the example above. Perhaps we could take the approach of doing the feature engineering in a single Layer subclass that takes in a dict of data. What do you think?

@fchollet Hi Francois. Thanks for the suggestion. Changes made to use a single feature layer by subclassing Layer class and using a dict of Input objects as input. Please let me if this is also what you had in mind.

**Review by fchollet — COMMENTED** [review]
Thanks for the update. A lot of the complexity here comes from the fact that you use a separate Input layer for each feature in the data, which isn't necessary if you use a Layer subclass. In addition, we should be showcasing Keras preprocessing layers.


I recommend something like:

```python
class FeaturePreprocessing(layers.Layer):
    def __init__(self):
        # Create preprocessing layers that will be needed for feature encoding / normalization / etc

    def adapt(self, dataset):
        # Split the dataset into individual feature datasets and use them to adapt the previously created  layers

    def call(self, data):
        # Preprocess the data dict with the previously created layers, then concatenate the features
        
```

Does that make sense? Perhaps a different dataset might be a better fit too, since we're going to want to do things like:

- Indexing a set of categorical string values
- Indexing a set of categorical int values
- Normalizing numerical features
- Hashing large categorical feature spaces
- etc.

**Comment by fernandonieuwveldt** [comment]
Hi @fchollet . Thanks for the suggestions. We now have one layer for feature preprocessing that utilises keras preprocessing layers. I have implemented the class you suggested. It contains multiple preprocessing layers and combinations of them.

Please let me know what you think. Is the dataset used fine for showcasing preprocessing layers? I was thinking to use a model from decision forest library instead neural network.

File: examples/structured_data/feature_preprocessing_layer.py [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+Keras by implementing a custom Keras layer for all feature preprocessing steps.This layer
+utilizes Keras preprocessing layers. For constructing our model we will use the Functional API.
+"""
+
+"""
+## Setup
+"""
+
+import tensorflow as tf
+from tensorflow.keras.layers import (
```
Comment (fchollet) [inline examples/structured_data/feature_preprocessing_layer.py:34]: Don't import individual layers, just `from tensorflow.keras import layers` and access them from `keras.layers`

File: examples/structured_data/feature_preprocessing_layer.py [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+labels = data_frame.pop("income_bracket")
+labels_binary = 1.0 * (labels == " >50K")
+
+print(f"Train dataset shape: {data_frame.shape}")
+
+"""
+## Split features into different feature groups
+"""
+
+NUMERIC_FEATURES = [
```
Comment (fchollet) [inline examples/structured_data/feature_preprocessing_layer.py:81]: NUMERICAL_FEATURES

File: examples/structured_data/feature_preprocessing_layer.py [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+CATEGORICAL_FEATURES = [
+    "workclass",
+    "marital_status",
+    "relationship",
+    "gender",
+    "education",
+    "occupation",
+]
+
+DISCRETIZE_FEATURES = ["age"]
```
Comment (fchollet) [inline examples/structured_data/feature_preprocessing_layer.py:98]: FEATURES_TO_DISCRETIZE

File: examples/structured_data/feature_preprocessing_layer.py [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+
+EMBEDDING_FEATURES = ["native_country"]
```
Comment (fchollet) [inline examples/structured_data/feature_preprocessing_layer.py:100]: FEATURES_TO_EMBED

File: examples/structured_data/feature_preprocessing_layer.py [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+
+class FeaturePreprocessor(tf.keras.models.Model):
+    def __init__(self, *args, **kwargs):
+        super(FeaturePreprocessor, self).__init__(*args, **kwargs)
+        # set preprocessing layers
+        self.normalizer = Normalization()
+        self.categoric_string_lookup = StringLookup()
+        self.integer_lookup = IntegerLookup(output_mode="binary")
+        self.discretizer = Discretization(num_bins=10)
+        self.embedding_string_lookup = StringLookup()
```
Comment (fchollet) [inline examples/structured_data/feature_preprocessing_layer.py:128]: You should use one layer per individual feature, don't reuse layers across multiple features

File: examples/structured_data/feature_preprocessing_layer.py [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+data_url = "https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data"
+
+data_frame = pd.read_csv(data_url, header=None, names=CSV_HEADER)
+labels = data_frame.pop("income_bracket")
+labels_binary = 1.0 * (labels == " >50K")
+
+print(f"Train dataset shape: {data_frame.shape}")
+
+"""
+## Split features into different feature groups
```
Comment (fchollet) [inline examples/structured_data/feature_preprocessing_layer.py:78]: There needs to be some explanation about why you chose to treat certain features as numerical, categorical, to discretize, etc.

File: examples/structured_data/feature_preprocessing_layer.py [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+model.compile(
+    loss=tf.keras.losses.BinaryCrossentropy(),
+    optimizer=tf.keras.optimizers.RMSprop(learning_rate=0.001),
+    metrics=[
+        tf.keras.metrics.BinaryAccuracy(name="accuracy"),
+        tf.keras.metrics.AUC(name="auc"),
+    ],
+)
+
+model.fit(data_dict, labels_binary.values, epochs=10, validation_split=0.25)
```
Comment (fchollet) [inline examples/structured_data/feature_preprocessing_layer.py:224]: There needs to be a section where we use the model to do inference on new data and we discuss test performance.

File: examples/structured_data/feature_preprocessing_layer.py [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+        )
+
+
+"""
+## Compile and fit model
+
+We now have a feature preprocessing layer. Lets put everything together and fit the model:
+"""
+
+inputs = {
```
Comment (fchollet) [inline examples/structured_data/feature_preprocessing_layer.py:189]: Again, let's process each feature independently.

File: examples/structured_data/feature_preprocessing_layer.py [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+    "discretize_features": tf.keras.Input(
+        shape=(len(DISCRETIZE_FEATURES),), dtype=tf.float32, name="discretize_layer"
+    ),
+    "embedding_features": tf.keras.Input(
+        shape=(len(EMBEDDING_FEATURES),), dtype=tf.string, name="embedding_layer"
+    ),
+}
+
+# update preprocessing layer states
+preprocessing_layer = FeaturePreprocessor()
```
Comment (fchollet) [inline examples/structured_data/feature_preprocessing_layer.py:205]: Let's use the common pattern where during training we `map()` the preprocessor into the tf.data pipeline, and at test time, we create a model that includes the preprocessor.

e.g.
```
input_dict = ...
preprocessed_features = preprocessor(input_dict)
x = layers.Dense(64, activation="relu")(preprocessed_features)
outputs = ...

training_model = keras.Model(inputs=preprocessed_features, outputs=outputs)
inference_model = keras.Model(inputs=input_dict, outputs=outputs)
```

Also let's explain that we do this because we want to do asynchronous parallel preprocessing at training time.

**Review by fchollet — CHANGES_REQUESTED** [review]
Thanks for the update!

**Comment by fernandonieuwveldt** [comment]
> 

Hi. Let me give it another go and than we can see if this will be a good addition to the website.

File: examples/structured_data/feature_preprocessing_layer.py (hunk 10 above)
Comment (fernandonieuwveldt) [reply examples/structured_data/feature_preprocessing_layer.py:282]: I have implemented this pattern on the changes.

Comment (fernandonieuwveldt) [reply examples/structured_data/feature_preprocessing_layer.py:282]: Hi @fchollet . This is quite an old PR and I have updated it now with all the suggestions and changes requested. Apologies that I only updated it now. Please have a look and inform me if we can still work on this.

**Comment by fernandonieuwveldt** [comment]
> > 
> 
> Hi. Let me give it another go and than we can see if this will be a good addition to the website.



> Hi @fernandonieuwveldt, thanks again for this PR. Are you planning to make the requested changes? Let us know if you're still working on this. Otherwise we'll close the request. Thanks!

Hi. I made the requested changes. Hope we can still work further on this.

**Comment by fernandonieuwveldt** [comment]
@fchollet @pcoet Let me know if the changes look good. All the requested changes have now been implemented.
